File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/sublist/sublist.go
package sublist

// Relation type is defined in relations.go file.

func Sublist(l1, l2 []int) Relation {
	if equal(l1, l2) {
		return RelationEqual
	}
	if isSublist(l1, l2) {
		return RelationSublist
	}
	if isSublist(l2, l1) {
		return RelationSuperlist
	}
	return RelationUnequal
}

// equal checks if two slices have the same elements in the same order
func equal(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// isSublist checks if slice a is a sublist of slice b
func isSublist(a, b []int) bool {
	// An empty list is a sublist of any list
	if len(a) == 0 {
		return true
	}
	// If a is longer than b, it can't be a sublist
	if len(b) < len(a) {
		return false
	}
	
	// Check for each possible starting position in b
	for i := 0; i <= len(b)-len(a); i++ {
		match := true
		for j := 0; j < len(a); j++ {
			if b[i+j] != a[j] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

File: src/2025-02-27-deepclaude37-rust/cpp/exercises/practice/sublist/sublist.cpp
#include "sublist.h"

namespace sublist {

namespace {
    // Helper function to check if one list is a sublist of another
    bool is_contained(const std::vector<int>& sub, const std::vector<int>& super) {
        // An empty list is a sublist of any list
        if (sub.empty()) {
            return true;
        }
        
        // If the potential sublist is longer than the superlist, it can't be a sublist
        if (sub.size() > super.size()) {
            return false;
        }
        
        // Check all possible starting positions in the superlist
        for (size_t i = 0; i <= super.size() - sub.size(); ++i) {
            bool match = true;
            
            // Check if the subsequence starting at position i matches the sublist
            for (size_t j = 0; j < sub.size(); ++j) {
                if (super[i + j] != sub[j]) {
                    match = false;
                    break;
                }
            }
            
            if (match) {
                return true;
            }
        }
        
        return false;
    }
}

List_comparison sublist(const std::vector<int>& list_a, const std::vector<int>& list_b) {
    // Check for equality first
    if (list_a == list_b) {
        return List_comparison::equal;
    }
    
    // Check if A is a sublist of B
    if (is_contained(list_a, list_b)) {
        return List_comparison::sublist;
    }
    
    // Check if B is a sublist of A (making A a superlist of B)
    if (is_contained(list_b, list_a)) {
        return List_comparison::superlist;
    }
    
    // If neither is a sublist of the other, they are unequal
    return List_comparison::unequal;
}

}  // namespace sublist
